lib/helpers/processConnectionOptions.js

```diff
@@ -0,0 +1,93 @@
+const clone = require('./clone');
+const MongooseError = require('../error/index');
+
+function processConnectionOptions(uri, options) {
+    let opts = options ? options : {};
+    const readPreference = opts.readPreference
+        ? opts.readPreference
+        : getUriReadPreference(uri);
+
+    const resolvedOpts = readPreference
+        ? resolveOptsConflicts(readPreference, opts)
+        : opts;
+
+    return clone(resolvedOpts);
+}
+
+function resolveOptsConflicts(pref, opts) {
+    const configIdx = opts.config && opts.config.autoIndex;
+    const { autoCreate, autoIndex, useCreateIndex } = opts;
+
+    // don't silently override user-provided indexing options
+    if (configIdx || autoCreate || autoIndex || useCreateIndex) {
+        const conflictLevel = getWarningLevel(pref);
+
+        if (conflictLevel === warnLevels.FAIL) {
+            throwReadPreferenceError();
+        } else if (conflictLevel === warnLevels.WARN) {
```

Comment:
I'd recommend getting rid of the `WARN` logic entirely, unless we add an extra option to explicitly suppress the warning.

Reply:
No problem.  It seems like we could eliminate the warning behavior in 1 of 4 ways:
1.  Permit the user to set index-creation options for connections that *could* connect to a non-primary and fail somewhat cryptically.
2. Throw for `nearest` and `primaryPreferred` in the same way as `secondary` and `secondaryPreferred`.
3. Override the user's preference and default to false anyway.

I *think* I prefer 2, but after some consideration I'm not sure that silently overriding is as bad as I first thought.  Maybe a user who's savvy enough to notice the override is also savvy enough not to set indexing options on a secondary connection, so we'd end up only throwing for less-familiar users who might prefer a connection that Just Works(TM).  In that case we could cut out the error path entirely, which would simplify the code a fair bit.  Do you have a preference?

Reply:
I prefer (1) actually. Index builds only happen once when the app starts up, so it is reasonable to connect using `primaryPreferred` for durability with the assumption that, when the app starts, the primary will be fine, and you're ok with the index build failing if there isn't a primary.

Reply:
Cool, sounds good.  I've got all these changes ready to go.  It looks like 6.0 has moved on since I forked it--would you like me to merge/rebase those changes into this PR, or just fix the commented issues?  Thanks!

Reply:
I went ahead and implemented all the changes from this PR/comments against the latest commit of 6.0, and rolled the changes up into a single commit.  The warn logic and all the unnecessary formatting diff should be gone.  The tests that are failing on this commit are failing on the tip commit of 6.0, so I don't think I introduced any new issues.  Please let me know if there's anything else you'd like changed.  Thanks!